### ISSUE TITLE
Suggestion of Making a CLI for FastApi

### ISSUE THREAD

**Issue by MedAmineFouzai** [tiangolo/fastapi#1568]
actually there is no issue i contacted mister Sebastián to propose for him an idea about making a CLI for the framework  to create a project structure  and managing our projects so that we developers have  start point that facilitate the hard work of making an architecture to the project therefore stays flexible for  expanding the project and adding more features and  automate the boring stuff like creating CRUD from the ORM that will make the project 1000% time faster than before .
locking dependencies by creating virtual environment add README file for documentation the project etc... 

i really hope some one help me making this become true because i dont know how to make a good cli but i managed to create something for tornado i called it  " cycloneCLI " for creating my own projects  then i found something called [torn](https://pypi.org/project/torn/) for tornado framework the problem is its a very very good structure and a starter project template that no one uses it it may be because the framework is starting to die but it helped me a lot to understand my needs better and work on different parts  of my project with more efficiency and good quality code so why we dont make fastapi better . and thank you
.
"med amine fouzai aka John constantine"

**Comment by dbanty** [comment]
Is the goal just initial project generation? If so, there are quite a few cookiecutter templates available to create projects with varying levels of features. There are some official ones at https://fastapi.tiangolo.com/project-generation/ and a bunch of third party ones under https://fastapi.tiangolo.com/external-links/#projects . Cookiecutter is a pretty standard way to go for this sort of thing.

If the goal is not project generation, what sort of other features are you looking for?

**Comment by dbanty** [comment]
On a second read, I see there are two pieces mentioned that project generation wouldn’t get you.

1. Locking dependencies and virtual environment- I suggest Poetry for this though there are some other options.

2. Generating CRUD functions from ORM models. This is an interesting idea but kind of a giant topic.  You could generate things for super simple access patterns (e.g. get by primary key) but most ORMs already have a clean way to do the simple things. You have any examples of the sort of things you’d like to generate?

**Comment by MedAmineFouzai** [comment]
mr dbanty  thank you for responding  ,what i'm looking is a start point way that all the developers using fastapi can start with for example the django-admin for creating apps and the rest-framework for REST APIS what i wont is smothing like that to enhance fastapi productivity twice the time i tried a lot of simller frameworks and i dont actully wont to see a third party module that generate my stuff i wont the core of the framework to handle this stuff thats why it called a framework 
and at the same time it will be awesome to see a new fresh tech having a lot of promised feature
what i'm saying is the python devs don't have a lot of features in the frameworks side this is a first time where i see a framework that supports graphql from the start its true that it uses graphene as a thired party module but its better than nothing also this makes me thnik twice revew my dev stack .

**Comment by MedAmineFouzai** [comment]
> On a second read, I see there are two pieces mentioned that project generation wouldn’t get you.
> 
> 1. Locking dependencies and virtual environment- I suggest Poetry for this though there are some other options.
> 2. Generating CRUD functions from ORM models. This is an interesting idea but kind of a giant topic.  You could generate things for super simple access patterns (e.g. get by primary key) but most ORMs already have a clean way to do the simple things. You have any examples of the sort of things you’d like to generate?

its true i use Anaconda to handle my dependencies and virtual environment but its able to use virtualenv inside the framework to handel it and lock dependencies ,and i  agree with  you in the Generating CRUD functions from ORM models it may be  kind of a giant topic but its doable .
look at symfony(php) for example it it has doctrine for making controllers out of view and if you see this as i do believe me the amount of speed gained from this uncredible projects will end in less than 10 lines of code

**Comment by tiangolo** [comment]
Yeah, as @dbanty says, you can check the project generators, that would be the first and simplest use case. :rocket: 

Also, about a CLI to generate FastAPI parts, FastAPI is designed to reduce the boilerplate and code duplication as much as possible. So, most or all of the code you write using FastAPI is actually related to your own app, not to required boilerplate, like what is the path, the HTTP operation, the request body, etc. A CLI tool to generate that would probably have to take a lot of arguments in the CLI to generate the same things in code. So, it would probably end up meaning that users need to learn a new CLI app, with a lot of CLI arguments and options, that would generate very little code, mostly just copying the data passed in those CLI arguments and options. But in that case, it's actually simpler to just write it in your code directly, instead of passing through the extra layer of abstraction with a CLI.

On the other side, for something more complex, it would depend on each different use case, for example, it would probably depend on each specific ORM, database, deployment setup, etc. So it wouldn't belong to FastAPI itself. But of course, if you need something more complex than the Cookiecutter project generators for your own projects, you can always build your own CLI. For example, with Typer: https://github.com/tiangolo/typer

**Comment by MedAmineFouzai** [comment]
thnak you mister Sebastián Ramírez for your reply ,i 'll go with your 
Suggestion  and i  will create a thierd party libary that generates a fastapi project for the community using typer  for any one needs a simple and helpfull structure to begin with and i will add other features in the way  and thank you again.

**Comment by ycd** [comment]
Hey, @MedAmineFouzai i don't know what you have done, but just randomly saw this issue. I'm working on [manage-fastapi](https://github.com/ycd/manage-fastapi)  i would be happy to hear your thoughts and ideas :)
